### PR TITLE
Polygonize improvements

### PR DESCRIPTION
**Pull request by cholmes**
A few improvements mentioned in #27: 
 * added a `min_size` parameter to filter out all the little slivers that usually aren't real fields. Default value of 500, which is about 5 pixels of sentinel 2. Requires an area calculation with reprojection which does slow things down a good bit, but seems like it may be worth it as the results seem a good bit better.
 * Added an 'area' parameter. More of a side effect of the filtering, but will be useful. 
 * Set a default simplification value of '15', as it seems to have pretty reasonable results.

Can see the difference, some places there's lots of these little ones that are just like a pixel big.

slivers-ftw

Would be good to also figure out the places where there's a one or two pixel hole in the polygon, seems like there should be something in shapely for it but I don't know it off hand. It does feel like a more sophisticated polygonization algorithm could do better, but it's pretty cool that we can get some fairly reasonable results just from rasterio.

I haven't done extensive testing on the performance of the area calculation, but when I did first add it in I definitely noticed. But overall it's not egregiously long, so may be worth it. If it was just to add an attribute that not everyone cares about we could make it an option. But filtering out these tiny polygons seems pretty useful and for that I'm pretty sure you need the area calculation w/ reprojection (maybe could do a faster filtering out pass without projecting to 6933? But probably not worth the effort), so I lean towards just doing it for everything.

@calebrob6 - I still have your warning about values greater than 1 for simplification in (but commented out). Curious when that would happen, since all the input will be sentinel 2, in UTM, so seems like for now we could count on the units always being meters? I do think it could be interesting to evolve this to be a more general 'simplification of satellite imagery inference' thing, but that feels not immediate, and seems reasonable to just assume sentinel 2, and then if we start trying other imagery sources we can adjust as needed.

### CODE REVIEW

**Comment by cholmes** [comment]
This should have readme updates before merging... but I may just continue to work on it adding in more improvements from the list in #27

**Comment by calebrob6** [comment]
> Curious when that would happen, since all the input will be sentinel 2

Yeah it won't unless someone reprojects into EPSG:4326 (maybe downloading from GEE or something). I think it is safe to remove! (the check is an instinct from other projects where sometimes our inputs are EPSG:4326 and sometimes local UTM)

File: src/ftw_cli/inference.py [hunk 1]
```diff
@@ -243,40 +244,52 @@ def polygonize(input_fn, output_fn, simplify, overwrite):
...
                     new_transform = transform * Affine.translation(x, y)
                     mask_window = mask[y:y+polygonization_stride, x:x+polygonization_stride]
                     for geom, val in rasterio.features.shapes(mask_window, transform=new_transform):
                         if val == 1:
-
+                            geom = shapely.geometry.shape(geom)
                             if simplify is not None:
-                                geom = shapely.geometry.shape(geom)
                                 geom = geom.simplify(simplify)
+
```
Comment (calebrob6) [inline src/ftw_cli/inference.py:266]: From memory, I think if you want to add the option to get rid of interior holes you can do:

```
if close_interiors:
    geom = shapely.geometry.Polygon(geom.exterior)
```

**Comment by calebrob6** [comment]
> I haven't done extensive testing on the performance of the area calculation, but when I did first add it in I definitely noticed. But overall it's not egregiously long, so may be worth it.

If it gets excessive then I think we can switch to a pyproj based approach-- I think I benchmarked fiona's transform and it was way slower but I can't find that.

File: src/ftw_cli/inference.py (hunk 1 above)
Comment (calebrob6) [reply src/ftw_cli/inference.py:266]: Yep!

image

**Comment by hannah-rae** [comment]
@cholmes these changes LGTM! I agree it would be nice to find a way to fill the holes - though sometimes there are real holes in a field, it would be nice to have the option (just like the area filtering).

File: src/ftw_cli/inference.py (hunk 1 above)
Comment (cholmes) [reply src/ftw_cli/inference.py:266]: Oh cool. Though I was hoping for something more nuanced - imagining that we'd let people set a threshold for removing holes. Maybe that's overkill - I was just thinking of it having work like the command I added to remove the tiny bits, where you set a threshold. 

I don't think there's actually fields that really have holes in them, so generally it's the right thing to do. But I'm imagining a case where there's some really big holes based on the raster, and user would prefer to see how it went wrong and get a closer representation of what the algorithm did.

But this is probably good enough for a first pass - we can make it an option so there's control over whether to do it or not.

**Comment by cholmes** [comment]
>  I agree it would be nice to find a way to fill the holes - though sometimes there are real holes in a field, it would be nice to have the option (just like the area filtering).

Oh, didn't see this before [my comment](https://github.com/fieldsoftheworld/ftw-baselines/pull/31#discussion_r1817201474). Didn't realize there are valid 'holes in fields' use cases, but actually I guess that does make sense. I think I lean towards the on/off option that does Caleb's operation for now, and then ticket that we'd ideally like control just like in the 'simplify' case.

**Comment by cholmes** [comment]
> Update: I remade it https://gist.github.com/calebrob6/2a2705027764d17fdbeb2b7a4cf4887c -- pyproj.Transformer is atleast ~2x faster.

Oh cool, I'll try to incorporate the code.

> But also, if we're already assuming that the inputs are going to be in the correct UTM projection, then it is probably fine to not reproject to 6933 (or anything else) to compute areas.

Ah, good to know - I wasn't sure, and this bit me in the past. I do think eventually we may want to handle other projections, so might be good to keep the code around for our future selves or future others.

**Comment by cholmes** [comment]
You want to review this @m-mohr? I think we should just get it merged and not wait on the holes stuff - I made an issue (#39) for it.

Then you could likely take on #27 and that one.

File: src/ftw_cli/inference.py (hunk 1 above)
Comment (m-mohr) [reply src/ftw_cli/inference.py:285]: Added to #27

**Comment by m-mohr** [comment]
Merging for now, this part of the CLI needs further work anyway. O(n³) complexity is not ideal... :-)

**Pull request merged**